File: frontend/src/api/positions.js
import api from "./client";

export async function listPositions(statusFilter) {
  const { data } = await api.get("/positions", {
    params: statusFilter ? { status: statusFilter } : {},
  });
  return data;
}

export async function getPosition(id) {
  const { data } = await api.get(`/positions/${id}`);
  return data;
}

export async function createPosition(payload) {
  const { data } = await api.post("/positions", payload);
  return data;
}

export async function updatePosition(id, payload) {
  // PATCH accepts stop_loss, target, quantity, status - whatever the user sends
  const { data } = await api.patch(`/positions/${id}`, payload);
  return data;
}

export async function deletePosition(id) {
  const { data } = await api.delete(`/positions/${id}`);
  return data;
}

export async function closePosition(id, exitPrice) {
  const { data } = await api.post(`/positions/${id}/close`, { exit_price: exitPrice });
  return data;
}

File: backend/routes/positions.py
# Thin routes - each one just pulls the logged-in user + DB session
# and hands the real work to position_service.

import uuid
from typing import Optional
from fastapi import APIRouter, Depends, status, Query
from sqlalchemy.orm import Session

from core.database import get_db
from core.dependencies import get_current_user
from models.user import User
from models.position import PositionStatus
from schemas.position import PositionCreate, PositionUpdate, PositionOut, ClosePositionRequest
from services import position_service

router = APIRouter(prefix="/positions", tags=["positions"])


@router.post("", response_model=PositionOut, status_code=status.HTTP_201_CREATED)
def create_position(
    payload: PositionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return position_service.create_position(db, current_user, payload)


@router.get("", response_model=list[PositionOut])
def list_positions(
    status_filter: Optional[PositionStatus] = Query(None, alias="status"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    # ?status=OPEN or ?status=CLOSED - omit to get everything (used by
    # the Dashboard for open positions and Trade History for closed ones)
    return position_service.list_positions(db, current_user, status_filter)


@router.get("/{position_id}", response_model=PositionOut)
def get_position(
    position_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return position_service.get_position(db, current_user, position_id)


@router.patch("/{position_id}", response_model=PositionOut)
def update_position(
    position_id: uuid.UUID,
    payload: PositionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return position_service.update_position(db, current_user, position_id, payload)


@router.post("/{position_id}/close", response_model=PositionOut)
def close_position(
    position_id: uuid.UUID,
    payload: ClosePositionRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    # Records the exit price, computes realized P&L, marks CLOSED -
    # this is what feeds the Trade History page.
    return position_service.close_position(db, current_user, position_id, payload.exit_price)


@router.delete("/{position_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_position(
    position_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    position_service.delete_position(db, current_user, position_id)
